File: rust/var-core/tests/matcher_test.rs
//! Port of `MatcherTest.java` / `matcher.test.ts`.

use var_core::handler::Handler;
use var_core::matcher::{ResolvedSteps, find_hits, resolve_hits};
use var_core::offsets::{utf16_index, utf16_len, utf16_slice};
use var_core::registry::{Registry, add_step, create_registry};
use var_core::value::Value;

fn reg() -> Registry {
    let r = create_registry();
    let r = add_step(
        &r,
        "I have {int} cukes",
        "steps.ts",
        1,
        Handler::noop(),
        None,
    )
    .unwrap();
    add_step(&r, "I withdraw {int}", "steps.ts", 5, Handler::noop(), None).unwrap()
}

#[test]
fn find_hits_returns_no_hits_when_nothing_matches() {
    assert!(find_hits("hello world", &reg()).is_empty());
}

#[test]
fn find_hits_returns_one_hit_per_step_expression_that_matches() {
    let hits = find_hits("Given I have 5 cukes in my belly", &reg());
    assert_eq!(1, hits.len());
    assert_eq!("I have {int} cukes", hits[0].expression);
    assert_eq!(6, hits[0].match_start);
    assert_eq!(20, hits[0].match_end);
    assert_eq!(vec![Value::Int(5)], hits[0].args);
}

#[test]
fn find_hits_returns_multiple_hits_when_multiple_expressions_match_non_overlapping_ranges() {
    let hits = find_hits("I have 5 cukes and I withdraw 3", &reg());
    let exprs: Vec<String> = hits.iter().map(|h| h.expression.clone()).collect();
    assert_eq!(
        vec![
            "I have {int} cukes".to_string(),
            "I withdraw {int}".to_string()
        ],
        exprs
    );
}

#[test]
fn resolve_hits_picks_longest_leftmost_when_ranges_overlap() {
    let r = create_registry();
    let r = add_step(&r, "I have {int} cukes", "s.ts", 1, Handler::noop(), None).unwrap();
    let r = add_step(
        &r,
        "I have {int} cukes in my belly",
        "s.ts",
        2,
        Handler::noop(),
        None,
    )
    .unwrap();
    let result = resolve_hits(find_hits("I have 5 cukes in my belly", &r));
    let ResolvedSteps::Ok(steps) = result else {
        panic!("expected Ok")
    };
    assert_eq!(1, steps.len());
    assert_eq!("I have {int} cukes in my belly", steps[0].expression);
}

#[test]
fn resolve_hits_returns_ambiguous_when_same_start_and_same_length_match() {
    let r = create_registry();
    let r = add_step(&r, "I have {int} cukes", "s.ts", 1, Handler::noop(), None).unwrap();
    let r = add_step(&r, "I have {int} {word}", "s.ts", 2, Handler::noop(), None).unwrap();
    let result = resolve_hits(find_hits("I have 5 cukes", &r));
    let ResolvedSteps::Ambiguous(collisions) = result else {
        panic!("expected Ambiguous")
    };
    assert_eq!(1, collisions.len());
    assert_eq!(2, collisions[0].candidates.len());
}

#[test]
fn resolve_hits_returns_all_non_overlapping_hits_left_to_right() {
    let r = create_registry();
    let r = add_step(&r, "I have {int} cukes", "s.ts", 1, Handler::noop(), None).unwrap();
    let r = add_step(&r, "I withdraw {int}", "s.ts", 2, Handler::noop(), None).unwrap();
    let result = resolve_hits(find_hits("Given I have 5 cukes and I withdraw 3", &r));
    let ResolvedSteps::Ok(steps) = result else {
        panic!("expected Ok")
    };
    let exprs: Vec<String> = steps.iter().map(|h| h.expression.clone()).collect();
    assert_eq!(
        vec![
            "I have {int} cukes".to_string(),
            "I withdraw {int}".to_string()
        ],
        exprs
    );
}

#[test]
fn param_spans_use_utf16_offsets_across_an_astral_character_no_manual_conversion_needed() {
    let r = create_registry();
    let r = add_step(&r, "I like {string}", "s.ts", 1, Handler::noop(), None).unwrap();

    let sentence = "😀 I like \"tea\"";

    let hits = find_hits(sentence, &r);
    assert_eq!(1, hits.len());
    let hit = &hits[0];

    let expected_match_start = utf16_index(sentence, sentence.find("I like").unwrap());
    assert_eq!(expected_match_start, hit.match_start);
    assert_eq!(utf16_len(sentence), hit.match_end);

    assert_eq!(1, hit.param_spans.len());
    let span = hit.param_spans[0];
    let quote_open = utf16_index(sentence, sentence.find('"').unwrap());
    assert_eq!(quote_open, span.start);
    assert_eq!(utf16_len(sentence), span.end);
    assert_eq!("\"tea\"", utf16_slice(sentence, span.start, span.end));
    assert_eq!(vec![Value::from("tea")], hit.args);
}

// -----------------------------------------------------------------------------
// Custom parameter types with capture groups (Java CaptureGroupTransformer /
// Python parse(*groups) parity)
// -----------------------------------------------------------------------------

#[test]
fn a_custom_type_with_capture_groups_passes_each_group_to_parse() {
    use std::rc::Rc;
    use var_core::registry::define_parameter_type;
    let r = define_parameter_type(
        &create_registry(),
        "range",
        r"(\d+)-(\d+)",
        Rc::new(|groups: &[&str]| {
            Value::list(groups.iter().map(|g| Value::from(*g)).collect::<Vec<_>>())
        }),
    );
    let r = add_step(&r, "the range is {range}", "s.rs", 1, Handler::noop(), None).unwrap();
    let hits = find_hits("the range is 10-20", &r);
    assert_eq!(1, hits.len());
    assert_eq!(
        vec![Value::list(vec![Value::from("10"), Value::from("20")])],
        hits[0].args
    );
}

#[test]
fn a_custom_type_without_groups_still_receives_the_whole_match() {
    use std::rc::Rc;
    use var_core::registry::define_parameter_type;
    let r = define_parameter_type(
        &create_registry(),
        "airport",
        "[A-Z]{3}",
        Rc::new(|groups: &[&str]| {
            assert_eq!(1, groups.len());
            Value::from(groups[0].to_lowercase())
        }),
    );
    let r = add_step(&r, "I fly to {airport}", "s.rs", 1, Handler::noop(), None).unwrap();
    let hits = find_hits("I fly to LHR", &r);
    assert_eq!(vec![Value::from("lhr")], hits[0].args);
}

File: java/core/src/test/java/dev/varar/core/MatcherTest.java
package dev.varar.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import org.junit.jupiter.api.Test;

/** Translated from {@code var-core/tests/matcher.test.ts}. */
class MatcherTest {

    private static final Object NOOP_HANDLER = (Runnable) () -> {};

    private static Registry reg() {
        Registry r = Registry.createRegistry();
        r = Registry.addStep(r, "I have {int} cukes", "steps.ts", 1, NOOP_HANDLER, null);
        r = Registry.addStep(r, "I withdraw {int}", "steps.ts", 5, NOOP_HANDLER, null);
        return r;
    }

    @Test
    void findHitsReturnsNoHitsWhenNothingMatches() {
        assertEquals(List.of(), Matcher.findHits("hello world", reg()));
    }

    @Test
    void findHitsReturnsOneHitPerStepExpressionThatMatches() {
        List<Matcher.Hit> hits = Matcher.findHits("Given I have 5 cukes in my belly", reg());
        assertEquals(1, hits.size());
        assertEquals("I have {int} cukes", hits.get(0).expression());
        assertEquals(6, hits.get(0).matchStart());
        assertEquals(20, hits.get(0).matchEnd());
        assertEquals(List.of(5), hits.get(0).args());
    }

    @Test
    void findHitsReturnsMultipleHitsWhenMultipleExpressionsMatchNonOverlappingRanges() {
        List<Matcher.Hit> hits = Matcher.findHits("I have 5 cukes and I withdraw 3", reg());
        assertEquals(
                List.of("I have {int} cukes", "I withdraw {int}"),
                hits.stream().map(Matcher.Hit::expression).toList());
    }

    @Test
    void resolveHitsPicksLongestLeftmostWhenRangesOverlap() {
        Registry r = Registry.createRegistry();
        r = Registry.addStep(r, "I have {int} cukes", "s.ts", 1, NOOP_HANDLER, null);
        r = Registry.addStep(r, "I have {int} cukes in my belly", "s.ts", 2, NOOP_HANDLER, null);
        List<Matcher.Hit> hits = Matcher.findHits("I have 5 cukes in my belly", r);
        Matcher.ResolvedSteps result = Matcher.resolveHits(hits);

        Matcher.Ok ok = assertInstanceOf(Matcher.Ok.class, result);
        assertEquals(1, ok.steps().size());
        assertEquals("I have {int} cukes in my belly", ok.steps().get(0).expression());
    }

    @Test
    void resolveHitsReturnsAmbiguousWhenSameStartAndSameLengthMatch() {
        Registry r = Registry.createRegistry();
        r = Registry.addStep(r, "I have {int} cukes", "s.ts", 1, NOOP_HANDLER, null);
        r = Registry.addStep(r, "I have {int} {word}", "s.ts", 2, NOOP_HANDLER, null);
        List<Matcher.Hit> hits = Matcher.findHits("I have 5 cukes", r);
        Matcher.ResolvedSteps result = Matcher.resolveHits(hits);

        Matcher.Ambiguous ambiguous = assertInstanceOf(Matcher.Ambiguous.class, result);
        assertEquals(1, ambiguous.collisions().size());
        assertEquals(2, ambiguous.collisions().get(0).candidates().size());
    }

    @Test
    void resolveHitsReturnsAllNonOverlappingHitsLeftToRight() {
        Registry r = Registry.createRegistry();
        r = Registry.addStep(r, "I have {int} cukes", "s.ts", 1, NOOP_HANDLER, null);
        r = Registry.addStep(r, "I withdraw {int}", "s.ts", 2, NOOP_HANDLER, null);
        List<Matcher.Hit> hits = Matcher.findHits("Given I have 5 cukes and I withdraw 3", r);
        Matcher.ResolvedSteps result = Matcher.resolveHits(hits);

        Matcher.Ok ok = assertInstanceOf(Matcher.Ok.class, result);
        assertEquals(
                List.of("I have {int} cukes", "I withdraw {int}"),
                ok.steps().stream().map(Matcher.Hit::expression).toList());
    }

    /**
     * Empirical UTF-16-offset verification (see {@code Matcher}'s class Javadoc and Task 14's
     * brief): places an astral character (an emoji outside the BMP — a surrogate pair, 2 {@code
     * char}s but 1 Unicode code point) before a captured {@code {string}} parameter, then checks
     * {@code matchStart}/{@code paramSpans} against offsets computed purely with {@code
     * String#indexOf}/{@code String#length} (which are UTF-16-code-unit-native in Java). If {@code
     * cucumber-expressions}'s {@code Group.getStart()/getEnd()} reported Unicode code-point offsets
     * instead (as raw {@code re}/{@code cucumber-expressions} do in the Python port, forcing {@code
     * to_utf16_offset}), every assertion below involving an offset *after* the emoji would be off
     * by exactly one and this test would fail.
     */
    @Test
    void paramSpansUseUtf16OffsetsAcrossAnAstralCharacterNoManualConversionNeeded() {
        Registry r = Registry.createRegistry();
        r = Registry.addStep(r, "I like {string}", "s.ts", 1, NOOP_HANDLER, null);

        String sentence = "😀 I like \"tea\""; // 😀 then " I like \"tea\""

        // Sanity check: this string really does contain an astral character, i.e. its
        // UTF-16 char count and Unicode code-point count differ — otherwise this test
        // would prove nothing.
        assertTrue(Character.isHighSurrogate(sentence.charAt(0)));
        assertEquals(sentence.length() - 1, sentence.codePointCount(0, sentence.length()));

        List<Matcher.Hit> hits = Matcher.findHits(sentence, r);
        assertEquals(1, hits.size());
        Matcher.Hit hit = hits.get(0);

        int expectedMatchStart = sentence.indexOf("I like");
        assertEquals(expectedMatchStart, hit.matchStart());
        assertEquals(sentence.length(), hit.matchEnd());

        assertEquals(1, hit.paramSpans().size());
        Matcher.ParamSpan span = hit.paramSpans().get(0);
        // cucumber-expressions' {string} capture group spans the whole quoted token
        // (quotes included) — the outer wrapping group compiled per-parameter, as
        // confirmed by the offsets below. The dequoted value ("tea", asserted via
        // hit.args() below) comes from a separate, inner group used only for
        // ParameterType#transform, not for this span.
        int quoteOpen = sentence.indexOf('"');
        int expectedParamStart = quoteOpen;
        int expectedParamEnd = sentence.length();
        assertEquals(expectedParamStart, span.start());
        assertEquals(expectedParamEnd, span.end());
        assertEquals("\"tea\"", sentence.substring(span.start(), span.end()));
        assertEquals(List.of("tea"), hit.args());
    }

    @Test
    void builtinEmphMatchesItalicEmphasisAndStripsTheDelimiters() {
        Registry r = Registry.addStep(Registry.createRegistry(), "I mention {emph}", "s.ts", 1, NOOP_HANDLER, null);
        List<Matcher.Hit> hits = Matcher.findHits("I mention *Emma*.", r);
        assertEquals(1, hits.size());
        assertEquals("I mention {emph}", hits.get(0).expression());
        // The built-in {emph} strips the outermost delimiter pair, passing the inner text.
        assertEquals(List.of("Emma"), hits.get(0).args());
    }

    @Test
    void builtinEmphMatchesBoldEmphasisAndStripsTheDelimiters() {
        Registry r = Registry.addStep(Registry.createRegistry(), "I mention {emph}", "s.ts", 1, NOOP_HANDLER, null);
        List<Matcher.Hit> hits = Matcher.findHits("I mention **Emma**.", r);
        assertEquals(1, hits.size());
        assertEquals(List.of("Emma"), hits.get(0).args());
    }
}
